qualtran/bloqs/arithmetic/conversions/sign_extension.py

```diff
@@ -0,0 +1,193 @@
+        return {(MultiTargetCNOT(self.extend_bitsize), 1)}
+
+    def on_classical_vals(self, x: 'ClassicalValT') -> dict[str, 'ClassicalValT']:
+        return {'y': x}
+
+
+@bloq_example
+def _sign_extend() -> SignExtend:
+    from qualtran import QInt
+
+    sign_extend = SignExtend(QInt(8), QInt(16))
+    return sign_extend
+
+
+_SIGN_EXTEND_DOC = BloqDocSpec(bloq_cls=SignExtend, examples=[_sign_extend])
+
+
+@frozen
+class SignTruncate(Bloq):
+    """Truncate a signed value to a smaller bitsize.
+
+    Useful to implement arithmetic operations with differing operand bitsizes.
+    A signed truncation xors the MSB (sign bit) into the bits to drop, and
+    deallocates them.
+
+    See :class:`SignExtend` for the adjoint operation.
+
+
+    Args:
+        inp_dtype: input data type.
+        out_dtype: output data type. must be same class as `inp_dtype`,
+                   and have smaller bitsize.
+
+    Registers:
+        x (LEFT): the input register of type `inp_dtype`
+        y (RIGHT): the output register of type `out_dtype`
+    """
+
+    inp_dtype: Union[QInt, QIntOnesComp]
+    out_dtype: Union[QInt, QIntOnesComp]
+
+    def __attrs_post_init__(self):
+        if not isinstance(self.inp_dtype, type(self.out_dtype)):
+            raise ValueError(
+                f"Expected same input and output base types, got: {self.inp_dtype}, {self.out_dtype}"
+            )
+
+        if not is_symbolic(self.truncate_bitsize) and self.truncate_bitsize <= 0:
+            raise ValueError(
+                f"input bitsize {self.inp_dtype.num_qubits} must be larger than "
+                f"output bitsize {self.out_dtype.num_qubits}"
+            )
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature(
+            [
+                Register('x', self.inp_dtype, side=Side.LEFT),
+                Register('y', self.out_dtype, side=Side.RIGHT),
+            ]
+        )
+
+    @cached_property
+    def truncate_bitsize(self):
+        return self.inp_dtype.num_qubits - self.out_dtype.num_qubits
+
+    def adjoint(self) -> 'SignExtend':
+        return SignExtend(self.out_dtype, self.inp_dtype)
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', x: 'Soquet') -> dict[str, 'SoquetT']:
+        xs = bb.split(x)
+        bits_to_drop, xs = xs[: self.truncate_bitsize], xs[self.truncate_bitsize :]
+
+        xs[0], bits_to_drop = bb.add(
+            MultiTargetCNOT(self.truncate_bitsize), control=xs[0], targets=bb.join(bits_to_drop)
+        )
+        bb.free(bits_to_drop)
+        x = bb.join(xs)
+
+        return {'y': x}
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> set['BloqCountT']:
+        return {(MultiTargetCNOT(self.truncate_bitsize), 1)}
+
+    def on_classical_vals(self, x: 'ClassicalValT') -> dict[str, 'ClassicalValT']:
+        bits = self.inp_dtype.to_bits(int(x))
+        y = self.out_dtype.from_bits(bits[self.truncate_bitsize :])
```

Comment:
I think we should assert here that all the dropped bits are 0 always (after xoring the MSB; or assert that all `bits` are equal to `y[0]`) -- as a sanity check that the adjoint is called on the correct input.

Reply:
added an assertion, and a unit test to see that it raises correctly